src/lib.rs

```diff
@@ -1,7 +1,1242 @@
+                dmax: Some([
+                    0x1a, 0x73, 0x0f, 0xeb, 0x00, 0x59, 0xcf, 0x1f, 0x5b, 0xde, 0xa8,
+                ]),
+            },
+        ];
+
+        let seed = [
+            0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23,
+            24, 25, 26, 27, 28, 29, 30, 31,
+        ];
+
+        let i1 = ChildIndex::NonHardened(1);
+        let i2h = ChildIndex::Hardened(2);
+        let i3 = ChildIndex::NonHardened(3);
+
+        let m = ExtendedSpendingKey::master(&seed);
+        let m_1 = m.derive_child(i1);
+        let m_1_2h = ExtendedSpendingKey::from_path(&m, &[i1, i2h]);
+        let m_1_2hv = ExtendedFullViewingKey::from(&m_1_2h);
+        let m_1_2hv_3 = m_1_2hv.derive_child(i3).unwrap();
+
+        let xfvks = [
+            ExtendedFullViewingKey::from(&m),
+            ExtendedFullViewingKey::from(&m_1),
+            ExtendedFullViewingKey::from(&m_1_2h),
+            m_1_2hv, // Appears twice so we can de-duplicate test code below
+            m_1_2hv_3,
+        ];
+        assert_eq!(test_vectors.len(), xfvks.len());
+
+        let xsks = [m, m_1, m_1_2h];
+
+        for j in 0..xsks.len() {
+            let xsk = &xsks[j];
+            let tv = &test_vectors[j];
+
+            let mut buf = [0; 32];
+            xsk.expsk.ask.into_repr().write_le(&mut buf[..]).unwrap();
+            assert_eq!(buf, tv.ask.unwrap());
+            xsk.expsk.nsk.into_repr().write_le(&mut buf[..]).unwrap();
+            assert_eq!(buf, tv.nsk.unwrap());
+
+            assert_eq!(xsk.expsk.ovk.0, tv.ovk);
+            assert_eq!(xsk.dk.0, tv.dk);
+            assert_eq!(xsk.chain_code.0, tv.c);
+
+            let mut ser = vec![];
+            xsk.write(&mut ser).unwrap();
+            assert_eq!(&ser[..], &tv.xsk.unwrap()[..]);
+        }
+
+        for j in 0..xfvks.len() {
+            let xfvk = &xfvks[j];
+            let tv = &test_vectors[j];
+
+            let mut buf = [0; 32];
+            xfvk.fvk.vk.ak.write(&mut buf[..]).unwrap();
+            assert_eq!(buf, tv.ak);
+            xfvk.fvk.vk.nk.write(&mut buf[..]).unwrap();
+            assert_eq!(buf, tv.nk);
+
+            assert_eq!(xfvk.fvk.ovk.0, tv.ovk);
+            assert_eq!(xfvk.dk.0, tv.dk);
+            assert_eq!(xfvk.chain_code.0, tv.c);
+
+            xfvk.fvk
+                .vk
+                .ivk()
+                .into_repr()
+                .write_le(&mut buf[..])
+                .unwrap();
+            assert_eq!(buf, tv.ivk);
+
+            let mut ser = vec![];
+            xfvk.write(&mut ser).unwrap();
+            assert_eq!(&ser[..], &tv.xfvk[..]);
+            assert_eq!(FVKFingerprint::from(&xfvk.fvk).0, tv.fp);
+
+            // d0
+            let mut di = DiversifierIndex::new();
+            match xfvk.dk.diversifier(di) {
+                Ok((l, d)) if l == di => assert_eq!(d.0, tv.d0.unwrap()),
+                Ok((_, _)) => assert!(tv.d0.is_none()),
+                Err(_) => panic!(),
```

Comment:
Because not all indices are valid diversifiers. `dk.diversifier()` takes an index `di`, and returns the index of the first valid diversifier occurring after that index (along with that diversifier). In this test, we actually want to know whether the specific diversifier index results in a valid diversifier or not; the way we test that is `l == di`. If that's true, we can then check we got the correct diversifier; if it's false, we check we were not actually expecting one.